main.cpp

```diff
@@ -1983,24 +1999,83 @@ uint32_t bootrom_table_lookup_rp2350(memory_access& access, uint16_t tag, uint16
     return raw_access.read_int(addr);
 }
 
-static rp2350_version_t get_rp2350_version(memory_access &raw_access) {
-    switch (get_rom_git_revision(raw_access)) {
-        case 0x312e22fa:
-            return rp2350_a2;
-        default:
-            return rp2350_unknown;
+static model_t determine_model(memory_access &raw_access) {
+    auto raw = raw_access.read_int(BOOTROM_MAGIC_ADDR);
+    auto magic = raw & 0xf0ffffff; // ignoring bootrom version
+    if (magic == BOOTROM_MAGIC_RP2040) {
+        return std::make_shared<model_rp2040>();
+    } else if (magic == BOOTROM_MAGIC_RP2350) {
+        uint32_t table_entry = raw_access.read_short(BOOTROM_MAGIC_ADDR + 4);
+        static_assert(ROM_END_RP2350 == 0x8000);
+        if (table_entry < 0x8000) {
+            return std::make_shared<model_rp2350>();
+        } else {
+            return std::make_shared<model_rp2350>(0x10000);
+        }
+    }
+    return models::unknown;
+}
+
+static inline bool is_transfer_aligned(uint32_t addr, const model_t& model) {
+    enum memory_type t = get_memory_type(addr, model);
+    return t != invalid && !(t == flash && addr & (PAGE_SIZE-1));
+}
+
+// this must be called after the right model is set on raw_access which is why it isn't
+// part of init_model
+static chip_revision_t determine_chip_revision(memory_access &raw_access) {
+    chip_revision_t chip_revision = unknown_revision;
+    model_t model = raw_access.get_model();
+    uint8_t rom_version;
+    raw_access.read_raw(0x13, rom_version);
+    if (model->chip() == rp2040) {
+        switch (rom_version) {
+            case 1:
+                chip_revision = rp2040_b0;
+                break;
+            case 2:
+                chip_revision = rp2040_b1;
+                break;
+            case 3:
+                chip_revision = rp2040_b2;
+                break;
+            default:
+                break;
+        };
+    } else if (model->chip() == rp2350) {
+        // switch (get_rom_git_revision_v2(raw_access)) {
+        //     default:
+        //         break;
+        // }
+        switch (rom_version) {
+            case 2:
+                chip_revision = rp2350_a2;
+                break;
+            case 3:
+                chip_revision = rp2350_a3;
+                break;
+            case 4:
+                chip_revision = rp2350_a4;
+                break;
+            default:
+                break;
+        };
     }
+    return chip_revision;
 }
 #if HAS_LIBUSB
 struct picoboot_memory_access : public memory_access {
-    model_t model = unknown; // must be initialized to something up front as it is referenced before it is set to its final value
     explicit picoboot_memory_access(picoboot::connection &connection) : connection(connection) {
-        model = get_model(*this);
+        // todo remove me temp hack so we can read a larger amount of ROM
+        model = std::make_shared<model_info>(unknown, "hack", 0x10000);
```

Comment:
Not required, as `determine_model` only reads `BOOTROM_MAGIC_ADDR` and `BOOTROM_MAGIC_ADDR + 4`, which is allowed by unknown model

```suggestion
```